### ISSUE TITLE
Decide the fate of custom site/project prefixing in the UserScript metadata block

### ISSUE THREAD

**Issue by Martii** [OpenUserJs/OpenUserJS.org#285]
In relation to greasemonkey/greasemonkey#1963 I see 3 possibilities so far over the past few days to address an impending conflict with everyones project.

``` javascript
// ==UserScript==
// @run-at             document-end
// @namespace          http://localhost.localdomain
// @name               RFC 2606§3 - @license and @licence (recovered) Unit Test
// @name:de-DE         RFC 2606§3 - @license und @licence (wiederhergestellt) Komponententest
// @name:fr            RFC 2606§3 - @license et @licence (récupéré) Test unitaire
// @description        Tests out non-unique keys
// @description:de-DE  Tests, nicht eindeutige Schlüssel
// @description:fr     Tests de clefs non unique
// @copyright          2007+, Marti Martz (http://userscripts.org/users/37004)
// @license            GPL version 3 or any later version; http://www.gnu.org/copyleft/gpl.html
// @version            0.0.0
// @include            http://www.example.com/*
// ==/UserScript==

// ==OpenUserJS==
// @author Marti
// @collaborator jerone
// @collaborator sizzle
// @collaborator Zren
// @hash 131a5eddc53f4018d6b5ee76d8543cb3cb4e50fa
// @somefutureidea somevalue
// ==/OpenUserJS==
```
* Has the disadvantage that any user.js engine may not read it. **EDIT**: ...but at the same time resolves any external conflicts and is in the spirit of optional collaboration instead of mutual collaboration.
* OUJS could store the **raw formatted metadata blocks text** instead of the single metadata block Object... and a migration routine put in place or applied. **EDIT**: We should do at least the latter part of this anyways for greasemonkey/greasemonkey#1963 support no matter what site/project prefixing we support. I highly recommend the former as well but depends on how we finalize the stored Object or not.

or

``` javascript
// ==UserScript==
// @run-at             document-end
// @namespace          http://localhost.localdomain
// @name               RFC 2606§3 - @license and @licence (recovered) Unit Test
// @name:de-DE         RFC 2606§3 - @license und @licence (wiederhergestellt) Komponententest
// @name:fr            RFC 2606§3 - @license et @licence (récupéré) Test unitaire
// @description        Tests out non-unique keys
// @description:de-DE  Tests, nicht eindeutige Schlüssel
// @description:fr     Tests de clefs non unique
// @copyright          2007+, Marti Martz (http://userscripts.org/users/37004)
// @license            GPL version 3 or any later version; http://www.gnu.org/copyleft/gpl.html
// @version            0.0.0
// @include            http://www.example.com/*
// @oujs.author Marti
// @oujs.collaborator jerone
// @oujs.collaborator sizzle
// @oujs.collaborator Zren
// @oujs.hash 131a5eddc53f4018d6b5ee76d8543cb3cb4e50fa
// @oujs.somefutureidea somevalue
// ==/UserScript==
```
* Using some other character. Since the damage is done with the dash operator may as well continue with destruction. Not backwards compatible. We also might run into this situation again if one of the user.js engines support the character we already do.

or

``` javascript
// ==UserScript==
// @run-at             document-end
// @namespace          http://localhost.localdomain
// @name               RFC 2606§3 - @license and @licence (recovered) Unit Test
// @name:de-DE         RFC 2606§3 - @license und @licence (wiederhergestellt) Komponententest
// @name:fr            RFC 2606§3 - @license et @licence (récupéré) Test unitaire
// @description        Tests out non-unique keys
// @description:de-DE  Tests, nicht eindeutige Schlüssel
// @description:fr     Tests de clefs non unique
// @copyright          2007+, Marti Martz (http://userscripts.org/users/37004)
// @license            GPL version 3 or any later version; http://www.gnu.org/copyleft/gpl.html
// @version            0.0.0
// @include            http://www.example.com/*
// @oujs:author: Marti
// @oujs:collaborator: jerone
// @oujs:collaborator: sizzle
// @oujs:collaborator: Zren
// @oujs:hash: 131a5eddc53f4018d6b5ee76d8543cb3cb4e50fa
// @oujs:somefutureidea: somevalue
// ==/UserScript==
```

This entails extra coding to maintain the `:` because we won't always know if a key is a site prefix or a locale... so must pad using an extra `:`. Padding will need to be maintained.

Everyone, especially the maintainers at @arantius @johan @gera2ld @derjanb @OpenUserJs/admin @OpenUserJs/owners @OpenUserJs/backend @OpenUserJs/frontend and @JasonBarnabe, is welcome to participate in this team effort meta discussion as usual.

Please let me know if any other ideas are better, worse, ineffective, etc. :)

Refs:
* #208
* #226
* #284
* greasemonkey/greasemonkey#1963
* greasemonkey/greasemonkey#1944
* greasemonkey/greasemonkey#1861
* https://github.com/Ventero/greasemonkey/commit/41c0f273ae7da2c169e4582363483d5184408463#commitcomment-7166075 and forward
* JasonBarnabe/greasyfork#215

**Comment by Martii** [comment]
After I get through rewriting oujs - Meta View *(Possibly the Alpha since it does pull script source for this to work properly)* it will be the Proof of Concept *(POC)* for *pegjs*.  So far everything appears cheesy. :) 

May strip off that `null` from the `locale` as we probably shouldn't be storing that in the DB... already did it in a test with a few more lines of code/grammar.

In the meantime are there any questions on the grammar files?

**Comment by Martii** [comment]
POC done... chatter welcome. :)

**Comment by Martii** [comment]
I believe that we can objectify the final object even further by keeping single object key names available for direct access instead of searching *(e.g. what the current parser is doing with Arrays)*... but `@name` and `@description` keys are special use cases where we would need a "find" method to test against the preferred locale and also I'd like to get "firstValueOf" and "lastValueOf" if possible... which we might need for non-unique keys on special occasions.

e.g. 
* if `meta.UserScript.name` doesn't exist then it's probably an Array of `@names` in which we would need to do a "find" abstract.
* if `meta.OpenUserJS.collaborator` doesn't exist it's probably an Array of `@collaborator`s in which we would need to do a "firstValueOf" and/or "lastValueOf" abstract.

or

We can just do "firstValueOf",  "lastValueOf" and "findValueOf" abstracts as the object sits in oujs - Meta View right now with the array of objects.

Any thoughts?

**Comment by Martii** [comment]
@sizzlemctwizzle 
If there is time and if possible... I'd like to create another issue ticket with sharing a JSON of this data

e.g. 
* https://openuserjs.org/meta/Marti/oujs_-_Meta_View.meta.js would optionally return just the barebones for updating in text
* https://openuserjs.org/meta/Marti/oujs_-_Meta_View.meta.json would return the object in JSON format

**Comment by Martii** [comment]
Slight hiccup *(fallout to use Anthony's terminology)* on migration which can't be reproduced yet with production itself... with existing UserScript `@author` key being in the wrong block but besides that appears to be good... **New scripts and modified scripts do not exhibit this behavior**... did **not** migrate the Graveyard `meta` values... that can be done on a slower time table and I'll have to reverse the migration from the old style to the new style... only reason to do this is if we have an "undo" command ever... but that also represents another challenge since the document record won't be maintained for Published date... yet another migration if we do it or do this feature.


@sizzlemctwizzle 
Thanks...need to keep blocking status for a day or two to see what issues might appear... I have tested this for about 60 days, give or take some changes here and there, so should be well.

**Comment by Martii** [comment]
Bingo... figured out the migration routine issue... all metadata is correct on production *(save for the Graveyard still)*... been doing this issue too long with too many variants of the code. :)

**Comment by Martii** [comment]
All dev script meta have been "extremely lazy" migrated to 0.2.0 of the project e.g. won't throw exceptions with current HEAD when accessing metadata
All dev removed *(Graveyard)* meta have been "extremely lazy" migrated to 0.2.0 of the project

All pro removed *(Graveyard)* meta have been "extremely lazy" migrated to 0.2.0 of the project

"extremely lazy" *(for me at least)* means that the old meta objects were transformed into the new meta objects unlike what I did with production scripts which the meta was regenerated from script source.

**Comment by Martii** [comment]
@sizzlemctwizzle 
I have encountered/seen **zero** issues... so this can be unblocked at your leisure. Went through all references and retested things *(removed mitigations, etc on other related issues/PRs)* 

Closure!!! :)

**Comment by Martii** [comment]
Closed by #717 and the plethora of comments here.

**Comment by Martii** [comment]
Reopened due to #729

**Comment by Martii** [comment]
As previously announced yesterday... taking server offline for migration hump... see everyone on the other side. :)

**Comment by Martii** [comment]
Back online... revalidating... on current master commit on production.

**Comment by Martii** [comment]
Graveyard is not complete yet... so leaving open until I get this done... this is a bit more complicated due to *mongoose* requirements of sub-Objects and the virtual "reason" I put in with `removerAutomated`... so I'll probably do this next week... again it's not as critical at this moment so a slower time table can happen with this. I do **not** need to take production offline to do these.

**Comment by Martii** [comment]
Graveyard should be complete.... also zapped that final script without source on sizzles account. Closing as redone.
